### PR TITLE
Avail plugin development

### PR DESCRIPTION
**Pull request by richATavail**
I added avail-plugin and a samples folder with a sample project. Just want feedback. If you all feel it is ready to merge, let’s do it. I’d like to have this stable by Thursday’s planned move to the base package ‘avail’

### CODE REVIEW

File: avail-plugin/src/main/kotlin/avail/plugin/AvailWorkbenchTask.kt [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+				"Run the Avail Workbench jar that was assembled " +
+					"in the parent task, `assembleAndRunWorkbench`."
+			workingDir = project.projectDir
+			jvmArgs(assembledVmOptions)
+			classpath = project.files(fullPathToFile)
+			println(buildString {
+				append("Launching: $fullPathToFile with VM options:")
+				assembledVmOptions.forEach {
+					append("\n\t • ")
+					append(it)
```
Comment (markATavail) [inline avail-plugin/src/main/kotlin/avail/plugin/AvailWorkbenchTask.kt:312]: Collapse to append("\n\t • $it").

File: samples/sample-project/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    mavenCentral()
+}
+
+dependencies {
+    // Can add an Avail library dependency as a jar available in one of the
+    // repositories listed in the repository section
+    // availLibrary("avail:example-lib:1.2.3")
+    testImplementation(kotlin("test"))
+}
+avail {
```
Comment (markATavail) [inline samples/sample-project/build.gradle.kts:53]: insert blank line after }

File: samples/sample-project/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+            // Add a module to this module package.
+            addModule("Configurations").apply {
+                // Specify module header for this module..
+                versions = listOf("Avail-1.6.0")
+                uses = listOf("Avail")
+                moduleHeaderCommentBody = customHeader
+            }
+            // Add a module package to this module package.
+            addModulePackage("Network").apply {
+                println("Settingup Network.avail")
```
Comment (markATavail) [inline samples/sample-project/build.gradle.kts:98]: Setting up

File: samples/sample-project/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        dependency("org.slf4j:slf4j-nop:2.0.0-alpha5")
+    }
+
+    // Add your own custom task to assemble and launch an Avail workbench.
+    val myWorkbenchTask by creating(AvailWorkbenchTask::class.java)
+    {
+        group = "My Tasks"
+        description = "My custom workbench build."
+        dependsOn(jar)
+        workbenchJarBaseName = "myCustomWorkBench"
```
Comment (markATavail) [inline samples/sample-project/build.gradle.kts:134]: lowercase bench

File: samples/sample-project/build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        rebuildWorkbenchJar = true
+        maximumJavaHeap = "6g"
+        workbenchLocalJarDependency("$buildDir/libs/sample-project.jar")
+        dependency("org.slf4j:slf4j-nop:2.0.0-alpha5")
+        root("my-avail-root", "$projectDir/avail/roots/my-avail-root")
+        root(
+            "avail",
+            "jar:$projectDir/avail/roots/avail-stdlib-1.6.0.20210910.181950.jar")
```
Comment (markATavail) [inline samples/sample-project/build.gradle.kts:142]: Isolate the version string, as an example of good practice for Avail users.
[I believe it's in Versions.availStripeVersion]

**Review by markATavail — APPROVED** [review]
Very nice work!

File: avail-plugin/README.md [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+            // Add a module inside the `App` module package.
+            addModule("Configurations").apply {
+            // The versions to list in the Avail header
+            versions = listOf("Avail-1.6.0")
+            // The modules to list in the uses section in the Avail header.
+            uses = listOf("Avail")
+            // Override the module header comment from 
+            // moduleHeaderCommentBodyFile
+            moduleHeaderCommentBody = customHeader
+        }
```
[the remark this answers is not in the document]
Comment (richATavail) [reply avail-plugin/README.md:349]: My branch shows as correctly indented and it does as well for the README viewer as presented in Github. I outdented and reindented for good measure.

File: avail-plugin/README.md [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+            uses = listOf("Avail")
+            moduleHeaderCommentBody = customHeader
+        }
+    }
+    
+    // Adds a root that will not be located in the `rootsDirectory`. 
+    root("imported-library", "jar:/Users/Some/Place/Else/imported-library.jar")
+    {
+        // Will be run as part of the initialize task
+        println("imported-library action has run!!!")
```
[the remark this answers is not in the document]
Comment (richATavail) [reply avail-plugin/README.md:377]: Again, can't really tell, but outdented and re-indented.

File: avail-plugin/README.md [hunk 8]
```diff
@@ -0,0 +1,486 @@
...
+            vmOption("-DavailDeveloper=true")
+	}
+    
+	test {
+		useJUnitPlatform()
+	}
+
+	withType<KotlinCompile>() {
+		kotlinOptions.jvmTarget = "16"
+	}
```
[the remark this answers is not in the document]
Comment (richATavail) [reply avail-plugin/README.md:445]: fixed. I did fight with IntelliJ quite a bit as it was being weird about this.

File: avail-plugin/buildSrc/.gitignore [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+/.gradle
+/.swp
+/build
+/stacks
+/modules.dot
+/out
+/debug
+/documentation/docs/src_docs
+/documentation/docs/index.md
+/documentation/site
```
[the remark this answers is not in the document]
Comment (richATavail) [reply avail-plugin/buildSrc/.gitignore:13]: I deleted that directory and removed the lines from .gitignore

File: avail-plugin/src/main/kotlin/avail/plugin/AvailExtension.kt [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+		/**
+		 * The directory location where the Avail roots exist.
+		 */
+		private const val defaultAvailRootsDirectory: String = "avail/roots"
+
+		/**
+		 * The directory location where the Avail roots repositories exist.
+		 */
+		private const val defaultRepositoryDirectory: String =
+			"avail/repositories"
```
[the remark this answers is not in the document]
Comment (richATavail) [reply avail-plugin/src/main/kotlin/avail/plugin/AvailExtension.kt:238]: They are used in the following manner:

```
var rootsDirectory: String =
	"${project.projectDir.absolutePath}/$defaultAvailRootsDirectory"
```
I did update it to be prefixed with a `.`, so `.avail/...`

I haven't introduced using ~./avail/repositories as a default. We can do that but I don't see any documentation indicating anything beyond `.avail/...`

**Comment by richATavail** [comment]
I fixed some bugs I found, cleaned up documentation, and extended the sample-project post @markATavail 's review.

File: avail-core/src/main/kotlin/com/avail/compiler/GlobalProgressReporter.kt [hunk 11]
```diff
@@ -36,7 +36,8 @@ package com.avail.compiler
  * The [compiler][AvailCompiler] notifies a `GlobalProgressReporter`
  * whenever a top-level statement is parsed unambiguously.
  *
- * The arguments are the total number of bytes of source being compiled and the
- * current number of bytes that have been compiled and executed.
+ * The arguments are:
+ * 1. the current number of bytes that have been compiled and executed
+ * 2. the total number of bytes of source being compiled and .
```
Comment (markATavail) [inline avail-core/src/main/kotlin/com/avail/compiler/GlobalProgressReporter.kt:41]: Remove " and " at end.

File: avail-plugin/build.gradle.kts [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+	}
+}
+
+plugins {
+	kotlin("jvm") version Versions.kotlin
+	`java-gradle-plugin`
+	`kotlin-dsl`
+	`maven-publish`
+}
+gradlePlugin {
```
Comment (markATavail) [inline avail-plugin/build.gradle.kts:51]: Insert blank line above.

**Review by markATavail — APPROVED** [review]

**Pull request merged**